### PR TITLE
Minor refactors in frontend JS

### PR DESCRIPTION
**Pull request by davwheat**
- If enabled in settings, use old mentions syntax when filling in from autocomplete
- Uses new core exports (throttle-debounce)
- Create new getMentionText function that assembles a mention string

### CODE REVIEW

**Comment by datitisev** [comment]
It's really hard to tell what each change is apart from TS and Prettier because it's all combined with TS changes (which the diff just shows deleted & new file) and Prettier changes.

**Comment by davwheat** [comment]
The only changes apart from the rewrite into TS are:

### Switching to throttle-debounce

https://github.com/flarum/mentions/blob/8b5ec2c99f9546fda58b59e7a6c74f4d574c3228/js/src/forum/addComposerAutocomplete.tsx#L21-L46

### Refactor the fetching and cleaning of display/user names to a function

https://github.com/flarum/mentions/blob/b164539ca5cc4701d956af112f46b200e62db6f3/js/src/forum/utils/getCleanDisplayName.ts#L1-L26

### Refactor the generation of mention text to a function

https://github.com/flarum/mentions/blob/b164539ca5cc4701d956af112f46b200e62db6f3/js/src/forum/utils/getMentionText.ts#L1-L37

File: js/.prettierrc.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  "printWidth": 150,
+  "singleQuote": true,
+  "tabWidth": 2,
```
[the remark this answers is not in the document]
Comment (tankerkiller125) [reply js/.prettierrc.json:4]: Nevermind I just double checked it is in fact 2.

**Comment by SychO9** [comment]
Not sure if this PR has any fixes related to the mentions bug, but can we extract that into a different minimal PR please ?

File: js/src/forum/utils/getCleanDisplayName.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * Fetches a user's username or display name.
+ *
+ * Chooses based on the format option set in the admin settings page.
+ *
+ * @param user An instance of the User model to fetch the username for
+ * @param useDisplayName If `true`, uses `user.displayName()`, otherwise, uses `user.username()`
+ */
+export default function cleanDisplayName(user, useDisplayName = true) {
```
Comment (askvortsov1) [inline js/src/forum/utils/getCleanDisplayName.js:18]: Why the rename?

Comment (davwheat) [reply js/src/forum/utils/getCleanDisplayName.js:18]: Prior to this, the function accepted a username input to clean up, but now the function fetches the username.

I didn't feel that cleanDisplayName was a good explanation, and that getCleanDisplayName was more clear.

Comment (askvortsov1) [reply js/src/forum/utils/getCleanDisplayName.js:18]: Let's change the file name then?

File: js/package.json [hunk 3]
```diff
@@ -2,12 +2,17 @@
...
+    "webpack-cli": "^4.7.0"
   },
   "scripts": {
     "dev": "webpack --mode development --watch",
     "build": "webpack --mode production"
+  },
+  "devDependencies": {
+    "@types/mithril": "^2.0.7",
+    "flarum": "^0.1.0-beta.16",
+    "flarum-tsconfig": "^0.1.0-beta.16.1"
```
Comment (SychO9) [inline js/package.json:16]: are these packages still on a beta versioning btw ?

File: js/src/forum/utils/getCleanDisplayName.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+/**
+ * Whether to use the old mentions format.
+ *
+ * `'@username'` or `'@"Display name"'`
+ */
+export const ShouldUseOldFormat = () => app.forum.data.attributes.allowUsernameMentionFormat || false;
+
+const GetDeletedUserText = () => app.translator.trans('core.lib.username.deleted_text');
```
Comment (SychO9) [inline js/src/forum/utils/getCleanDisplayName.js:8]: shouldn't non components not be capitalised ? might just be a react thing, but It's a good rule I believe.

**Review by SychO9 — COMMENTED** [review]
Looks good, haven't tested yet, we should aim to get this in for the next minor release.

a few minor comments.

File: js/package.json (hunk 3 above)
Comment (datitisev) [reply js/package.json:16]: `flarum` was used for typings, can be removed as we use `vendor` folder now - tsconfig is on v1.0.0

File: js/src/forum/utils/getCleanDisplayName.js (hunk 4 above)
Comment (datitisev) [reply js/src/forum/utils/getCleanDisplayName.js:8]: Yeah we use camel case for everything, this should continue to use it.

File: js/package.json [hunk 5]
```diff
@@ -1,13 +1,20 @@
 {
   "private": true,
   "name": "@flarum/mentions",
+  "prettier": "@flarum/prettier-config",
   "dependencies": {
-    "flarum-webpack-config": "0.1.0-beta.10",
-    "webpack": "^4.43.0",
-    "webpack-cli": "^3.3.11"
+    "flarum-webpack-config": "1.0.0",
```
Comment (SychO9) [inline js/package.json:6]:
```suggestion
    "flarum-webpack-config": "^1.0.0",
```

**Review by SychO9 — CHANGES_REQUESTED** [review]

File: js/src/forum/utils/getCleanDisplayName.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+/**
+ * Whether to use the old mentions format.
+ *
+ * `'@username'` or `'@"Display name"'`
+ */
+export const shouldUseOldFormat = () => app.forum.data.attributes.allowUsernameMentionFormat || false;
```
Comment (datitisev) [inline js/src/forum/utils/getCleanDisplayName.js:6]: Should this use `app.forum.attribute()`? Not sure why we'd access the data directly.

Comment (davwheat) [reply js/src/forum/utils/getCleanDisplayName.js]: I have no clue either... Fixed it now.

**Review by askvortsov1 — APPROVED** [review]
Left a few nits

File: js/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  // Use Flarum's tsconfig as a starting point
+  "extends": "flarum-tsconfig",
+  // This will match all .ts, .tsx, .d.ts, .js, .jsx files
+  "include": ["src/**/*"],
+  "compilerOptions": {
+    // This will output typings to `dist-typings`
+    "declarationDir": "./dist-typings",
+    "baseUrl": ".",
+    "paths": {
+      "flarum/*": ["../vendor/flarum/core/js/dist-typings/*"]
```
[the remark this answers is not in the document]
Comment (davwheat) [reply js/tsconfig.json:11]: Yes, but that didn't exist when I made this pr :)

**Review by SychO9 — APPROVED** [review]

**Pull request merged**